### PR TITLE
common: added MAV_CMD_CAN_FORWARD and CAN_FRAME

### PR DESCRIPTION
**Pull request by dagar**
- https://github.com/ArduPilot/mavlink/pull/244

### CODE REVIEW

File: message_definitions/v1.0/minimal.xml [hunk 1]
```diff
@@ -626,6 +626,9 @@
       <entry value="181" name="MAV_COMP_ID_BATTERY2">
         <description>Battery #2.</description>
       </entry>
+      <entry value="189" name="MAV_COMP_ID_MAVCAN">
```
Comment (hamishwillee) [inline message_definitions/v1.0/minimal.xml:629]: Can there be multiples? Will the component ever emit a heartbeat? If so you also need a MAV_TYPE

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -2455,6 +2455,16 @@
         <param index="7">User defined</param>
       </entry>
       <!-- END of user range (31000 to 31999) -->
+      <entry value="32000" name="MAV_CMD_CAN_FORWARD" hasLocation="false" isDestination="false">
+        <description>Request forwarding of CAN packets from the given CAN bus to this interface. CAN Frames are sent using CAN_FRAME messages</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:2459]: When you say interface, do you mean this system and component?

File: message_definitions/v1.0/common.xml [hunk 3]
```diff
@@ -6998,6 +7008,15 @@
       <field type="uint8_t" name="payload_length">Length of the data transported in payload</field>
       <field type="uint8_t[128]" name="payload">Variable length payload. The payload length is defined by payload_length. The entire content of this block is opaque unless you understand the encoding specified by payload_type.</field>
     </message>
+    <message id="386" name="CAN_FRAME">
+      <description>A forwarded CAN frame as requested by MAV_CMD_CAN_FORWARD.</description>
+      <field type="uint8_t" name="target_system">System ID.</field>
+      <field type="uint8_t" name="target_component">Component ID.</field>
+      <field type="uint8_t" name="bus">bus number</field>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:7015]:
```suggestion
      <field type="uint8_t" name="bus">Bus number</field>
```

File: message_definitions/v1.0/common.xml [hunk 4]
```diff
@@ -2455,6 +2455,16 @@
         <param index="7">User defined</param>
       </entry>
       <!-- END of user range (31000 to 31999) -->
+      <entry value="32000" name="MAV_CMD_CAN_FORWARD" hasLocation="false" isDestination="false">
+        <description>Request forwarding of CAN packets from the given CAN bus to this interface. CAN Frames are sent using CAN_FRAME messages</description>
+        <param index="1" label="bus">Bus number (0 to disable forwarding, 1 for first bus, 2 for 2nd bus, 3 for 3rd bus).</param>
+        <param index="2">Empty.</param>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:2461]: You could omit all these. By default the MAVLink spec says they will be generated as Reserved with default value of `0`.

**Review by hamishwillee — COMMENTED** [review]
Seems like this has been agreed by main stakeholders. @julianoes @auturgy Any comments on a MAVLink/ technical level?

@dagar @tridge 
- My understanding is that this is so a GCS can understand what is going on with CAN. So perhaps it will have some way to present this CAN traffic for users - is that right? The reason I ask is that it would be good to document the scope of this.
- Do we need to have some way to check whether the API is supported? E.g. a protocol bit?
- Is there a need to also inject CAN messages from a MAVLink component?

File: message_definitions/v1.0/common.xml (hunk 2 above)
Comment (julianoes) [reply message_definitions/v1.0/common.xml:2459]: I'd agree with component.

**Review by julianoes — COMMENTED** [review]
Would be nice to add that indeed.

**Comment by hamishwillee** [comment]
FYI, I won't merge this without an answer to https://github.com/mavlink/mavlink/pull/1792#pullrequestreview-876727435. Experience tells me that if we don't explain why we're doing something at least in the PR part, there is no way to get the information later.

**Comment by tridge** [comment]
> * My understanding is that this is so a GCS can understand what is going on with CAN. So perhaps it will have some way to present this CAN traffic for users - is that right? The reason I ask is that it would be good to document the scope of this.

it is already in the dronecan_gui_tool and in pydronecan. It is fully working now with the ArduPilot mavlink repo

> Do we need to have some way to check whether the API is supported? E.g. a protocol bit?

its a MAV_CMD, so it already has an UNSUPPORTED result

> Is there a need to also inject CAN messages from a MAVLink component?

that is in the implementation already, using the same CAN_FRAME messages

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@dagar I've added CANFD_FRAME and CAN_FILTER_MODIFY to this PR. Both are in the ArduPilot mavlink repo already. 
CANFD over mavlink has been tested, and is supported in pydronecan and dronecan_gui_tool

File: message_definitions/v1.0/minimal.xml (hunk 1 above)
Comment (tridge) [reply message_definitions/v1.0/minimal.xml:629]: these comp IDs are just a convention, not required. I've made the dronecan_gui_tool use it, but I expect the CAN_FRAME message will be implemented in all GCS over time.

File: message_definitions/v1.0/common.xml (hunk 4 above)
Comment (tridge) [reply message_definitions/v1.0/common.xml:2461]: personally I think it is clearer leaving them in

**Comment by dagar** [comment]
Are we good to go here?

**Comment by tridge** [comment]
> Are we good to go here?

I think so, but maybe you could approve?

**Comment by tridge** [comment]
i'll merge now .... seems pointless waiting on you to approve your own PR

**Pull request merged**